### PR TITLE
MuStruct binary serialization

### PR DESCRIPTION
**Pull request by esdoppio**
wip. pls don't merge

### CODE REVIEW

File: muschema/src/struct.ts [hunk 1]
```diff
@@ -39,23 +39,23 @@ export class MuStruct<StructSpec extends _SchemaDictionary>
         });
 
         const args:string[] = [];
-        const props:any[] = [];
+        const props:MuSchema<any>[] = [];
```
Comment (mikolalysenko) [inline muschema/src/struct.ts]: injected properties may not be schemas necessarily.  for example, long string constants might be good candidates for injection, or maybe other objects which are initialized in the containing scope.

File: muschema/src/struct.ts [hunk 2]
```diff
@@ -379,6 +381,41 @@ export class MuStruct<StructSpec extends _SchemaDictionary>
...
+                    byteLength += 4;
+                    break;
+                case 'float64':
+                    byteLength += 8;
+                    break;
+                case 'vector':
+                    byteLength += type.identity.byteLength;
+                    break;
+                case 'string':
+                    methods.getByteLength.push(`result+=4+Buffer.byteLength(x["${propName}"]);`);
```
Comment (mikolalysenko) [inline muschema/src/struct.ts]: Buffer may not be available in a browser environment.  Maybe we could just use `x["${propName}"].length * 4` as a crude upper bound?

File: muschema/src/struct.ts [hunk 3]
```diff
@@ -318,13 +320,13 @@ export class MuStruct<StructSpec extends _SchemaDictionary>
         // diffBinary subroutine
         const numProps = structProps.length;
         const trackerBytes = Math.ceil(numProps / 8);
-        methods.diffBinary.push(`var numPatch=0;var tracker=0;var trackerOffset = s.offset;s.grow(${trackerBytes});s.offset+=${trackerBytes};`);
+        methods.diffBinary.push(`var numPatch=0;var tracker=0;var trackerOffset = s.offset;s.grow(this.getByteLength(t)+${trackerBytes});s.offset+=${trackerBytes};`);
```
Comment (mikolalysenko) [inline muschema/src/struct.ts]: we should include `trackerBytes` in the byte length of the struct.

Comment (esdoppio) [reply muschema/src/struct.ts]: ah that's the plan, somehow missed that

File: muschema/src/struct.ts (hunk 2 above)
Comment (esdoppio) [reply muschema/src/struct.ts]: yes, will do. didn't occur to me

**Comment by esdoppio** [comment]
all set

**Pull request merged**